keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+  apiGroup: rbac.authorization.k8s.io
+```
+
+The webhook will fail the request if the opt-out annotation is present without the RBAC binding for the requestor.
+
+#### Race Condition
+When a namespace is created, there's a gap before all objects in the namespace
+templates are populated.
+During the gap, the users may have the access to create objects into namespaces.
+
+To avoid the situation, a few mechanisms can be used:
+
+##### Creating namespace through service
+
+Users don't have permission to create namespace directly using Kubernetes API.
+The namespace creation is performed using a custom service which runs under a
+service account with enough privilege to create namespaces and objects inside namespaces.
+The service will also grant the current user privilege to manage the namespace.
+
+This approach requires additional development effort and introduces non-standard way
+for namespace creation.
+
+##### Setup RBAC properly
+
+See above [Self-serviced namespace creation](#self-serviced-namespace-creation)
+for details of RBAC setup.
+
+This approach requires pre-setup work from cluster admin for ClusterRoleBinding to grant
+users/groups the privilege to create namespaces.
+Beyond that, it is the simplest and recommended approach.
+
+##### Use of Initializers
+
+The _NamespaceTemplate_ enforcement controller becomes a namespace initializer.
+A newly created namespace stays _uninitialized_ until the controller creates
+all predefined objects.
+
+However, the Initializers admission control only hides the namespace without
+block actual access in the namespace.
+As long as a user knows the name of the namespace, it's still possible to create
+objects inside.
+
+##### Use of ValidatingAdmissionWebhook
+
+An external _ValidatingAdmissionWebhook_ will block all create/update/patch/delete
+operations in an _uninitialized_ namespace, except those from the _NamespaceTemplate_
+enforcement controller (by whitelisting the service account running the controller).
+
+The webhook works similarly to _Initializers_. 
+The service account running the controller for initialization must be whitelisted.
+The recommended mechanism for whitelist is using RBAC:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-initialization
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    verbs: ["initialize"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: namespace-initializer
+roleRef:
+  kind: ClusterRole
+  name: namespace-initialization
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: ServiceAccount
+  name: "system:serviceaccount:kube-system:name"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Using external admission webhook increases the complexity, and also latency.
+As the webhook must be reliable all the time, it must be deployed in HA.
+It's recommended to create a built-in admission controller for namespace initialization.
```

Comment:
I don't think this statement is true.  I certainly don't think we should put this admission controller in code if it can be built with an external admission controller.

I'm on the fence about whether this feels like a part of Kube or not, but if this is a controller and a CRD and is plugging in to extension, I don't see anything that requires built in web hook admission yet, certainly not until it graduates from beta.

Reply:
I'm OK with building an external webhook. I will remove this statement.